editor/language_server/src/diagnostics.rs

```diff
@@ -0,0 +1,77 @@
+//! Module `diagnostics` converts slide diagnostics to LSP types.
+
+use libslide::diagnostics as s;
+use libslide::Span;
+use tower_lsp::lsp_types::*;
+
+pub fn convert_diagnostics(
+    diagnostics: &[s::Diagnostic],
+    provider: &str,
+    uri: &Url,
+    source: &str,
+) -> Vec<Diagnostic> {
+    diagnostics
+        .into_iter()
+        .map(|diagnostic| Diagnostic {
+            range: to_range(&diagnostic.span, source),
+            severity: Some(to_severity(&diagnostic.kind)),
+            code: Some(NumberOrString::String(diagnostic.code.to_string())),
+            source: Some(provider.to_string()),
+            message: flatten_diagnostic_msg(diagnostic),
+            related_information: Some(flatten_related(diagnostic, uri, source)),
+            tags: None,
+        })
+        .collect()
+}
+
+fn flatten_diagnostic_msg(diagnostic: &s::Diagnostic) -> String {
+    match &diagnostic.msg {
+        Some(msg) => format!("{} \\ {}", diagnostic.title, msg),
+        None => diagnostic.title.to_string(),
+    }
+}
+
+fn flatten_related(
+    diagnostic: &s::Diagnostic,
+    uri: &Url,
+    source: &str,
+) -> Vec<DiagnosticRelatedInformation> {
+    diagnostic
+        .associated_diagnostics
+        .iter()
+        .chain(diagnostic.unspanned_associated_diagnostics.iter())
+        .map(|ad| DiagnosticRelatedInformation {
+            location: Location::new(uri.clone(), to_range(&ad.span, source)),
+            message: ad.msg.to_string(),
+        })
+        .collect()
+}
+
+fn to_severity(diagnostic_kind: &s::DiagnosticKind) -> DiagnosticSeverity {
+    match diagnostic_kind {
+        s::DiagnosticKind::Error => DiagnosticSeverity::Error,
+        s::DiagnosticKind::Warning => DiagnosticSeverity::Warning,
+        s::DiagnosticKind::Note => DiagnosticSeverity::Information,
+        s::DiagnosticKind::Help => DiagnosticSeverity::Hint,
+    }
```

Comment:
this seems useless. why dont you just replace severity with kind in a duagnostic above. also to/from would be better if not

Reply:
I can't, I have to change the slide diagnostic type to work with the other diagnostic type, which expects this other enum